File: cpp/averageoflevelsinbinarytree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        queue<TreeNode*> que;
        vector<double> res;
        que.push(root);
        while (!que.empty()) {
            int num = que.size();
            double sum = 0;
            for (int i = 0; i < num; i++) {
                //cout << que.front()->val << endl;
                sum += que.front()->val/double(num);
                if (que.front()->left) que.push(que.front()->left);
                if (que.front()->right) que.push(que.front()->right);
                que.pop();
            }
            //cout << "sum:" <<sum << endl;
            res.push_back(sum);
        }
        return res;
    }
};

File: java/AverageOfLevelsInBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        Queue<TreeNode> que = new LinkedList<>();
        List<Double> res = new ArrayList<>();
        que.add(root);
        
        while (!que.isEmpty()) {
            Double sum = 0.0;
            int n = que.size();
            for (int i = 0; i < n; i++) {
                TreeNode tmp = que.poll();
                sum += tmp.val;
                if (tmp.left != null) que.add(tmp.left);
                if (tmp.right != null) que.add(tmp.right);
            }
            res.add(sum/n);
        }
        
        return res;
    }
}
